### ISSUE TITLE
VERBS;OFI_RXM: Data corruption when using multi-receive message using fi_sendmsg API

### ISSUE THREAD

**Issue by tonyzinger** [ofiwg/libfabric#5599]
When transferring data using the multi-receive flag on the fi_sendmsg API, a data corruption problem occurs.

The data corruption occurs in Rank 0 when it receives the data from Rank 5 during the 33 (0x21) transfer, which is the 9th fi_sendmsg request. Each fi_sendmsg request has 4 iovs, i.e. transfers.

The test does not fail all of the time but it does fail frequently enough.

I am attaching the debug output for Rank 0. I have tried to document in this debug output what was happening in the test.
[multi_recv_rank_0.txt](https://github.com/ofiwg/libfabric/files/4149852/multi_recv_rank_0.txt)

It appears that the received (target) CQs are being read in the order of the sends. However, the CQ that is sent when the multi-receive buffer is full, appears to come a couple of CQ entries too late. It appears to come after the next multi-receive buffer has received data for 2 additional fi_sendmsg requests.

The test receives the target CQ entries for transfers 0x21 and 0x25. The data for transfer 0x25 should fill the multi-receive buffer. At this time I would expect a CQ entry stating that the multi-receive buffer is full. However, I receive the target CQ entries for transfers 0x29 and 0x2d. Then I receive the CQ entry stating that the multi-receive buffer is full.

Before the data is verified, all of the expected send and received CQ entries have been process. Also, the interface counter is at the expected value. When the data is verified, the location for transfer 0x21 contains the data for transfer 0x25. And the next 15 data transfers are corrupted. These 16 transfers would be contained in 4 fi_sendmsg requests, i.e. 4 iovs per request.

This test includes the tx_attr and rx_attr msg_order flag: FI_ORDER_SAS, on the call to fi_getinfo().

This test still fails if the FI_TRANSMIT_COMPLETE or FI_DELIVERY_COMPLETE is set in the flags parameter on the fi_sendmsg and fi_recvmsg APIs.

**Comment by shefty** [comment]
I've confirmed that there are issues with the multi-recv handling in rxm.  I've merged several patches last week that help address this, and more are on the way.  (I know that there is an issue handling expected messages when posting a multi-recv buffer.)  At least one fix that was applied changed reposting multi-recv buffers to the _head_ of the receive queue, rather than the tail.

Can you test with the tip of master and report if the problem is still the same?

**Comment by tonyzinger** [comment]
I will get the latest master and try my test.

**Comment by tonyzinger** [comment]
I have tried using Libfabric from the tip of master.  However, I am seeing a different problem.
I executed my test case using 4 ranks in an All-to-All communication pattern sending 12 transfers.
Rank 0 did not receive all of the messages/data from rank 2.  It received the first 4 transfers. But
transfer 5 through 12 did not arrive.  Also, for rank 0 I did not receive the corresponding target CQ
events from rank 2.  Since all of the data did not arrive, those bytes are zero filled.  Whereas,
previously, all of the data arrived but it landed out of order.
Rank 0 received all of the data from rank 1 and rank 3 correctly.
Rank 1, rank 2 and rank 3 received all of their data from the other ranks correctly.
